src/training/cifar_eval.py

```diff
@@ -0,0 +1,90 @@
+import argparse
+import os
+
+import numpy as np
+import torch
+from torch.utils.data import DataLoader
+from datasets import load_dataset
+from joblib import Parallel, delayed
+from PIL import Image
+from sklearn.linear_model import LogisticRegression
+from vision_eval import mapper
+
+import open_clip
+
+
+def resize(imgs, size=(24, 24)):
+    if type(imgs['img']) == Image.Image:
+        imgs['img_resized'] = imgs['img'].resize(size)
+        return imgs
+    elif type(imgs['img']) == list:
+        imgs['img_resized'] = [img.resize(size) for img in imgs['img']]
+        return imgs
+
+def encode_batch(model, dataloader):
+    model.eval()
+    embeddings = []
+    with torch.no_grad():
+        for batch in dataloader:
+            embeddings.append(model.encode_image(batch))
+    return torch.cat(embeddings, dim=0)
+
+def preprocess_batch(dict, preprocessor, batch_size=64):
+    prep = []
+    for i in range(0, len(dict)):
+        prep.append(preprocessor(dict[i]['img_resized']).unsqueeze(0))
+    
+    return torch.cat(prep, dim=0)
+
+def generate_embeddings(args):
+    cifar = load_dataset("cifar10")
```

Comment:
Would recommend the following design pattern:

- create a Dataset object which handles reading from the huggingface dataset, resizing, and preprocessing the images
- when generating the embeddings, use a batched DataLoader to yield batches from the dataset object (they should already be resized/preprocessed) and call `encode_image` on this batch to produce the embeddings.
- You can probably take the embeddings off the accelerator if you're running out of memory (using `.cpu()`). You also may want to wrap calls to `encode_image` with `torch.no_grad()` to ensure PyTorch doesn't trace the activations (which burns a lot of memory)
- using the saved embeddings from the entire dataset, use sklearn to learn a classifier.

Reply:
agree on all except torch is already using the CPU since I don't have any GPUs atm